mlflow/gemini/chat.py

```diff
@@ -0,0 +1,188 @@
+            k: convert_gemini_param_property_to_mlflow_param_property(v)
+            for k, v in function_params.properties.items()
+        },
+        required=function_params.required,
+    )
+
+
+def convert_gemini_func_to_mlflow_chat_tool(
+    function_def: "genai.protos.FunctionDeclaration",
+) -> ChatTool:
+    """
+    Convert Gemini function definition into MLflow's standard format (OpenAI compatible).
+    Ref: https://ai.google.dev/gemini-api/docs/function-calling
+
+    Args:
+        function_def: A genai.protos.FunctionDeclaration object representing a function definition.
+
+    Returns:
+        ChatTool: MLflow's standard tool definition object.
+    """
+    return ChatTool(
+        type="function",
+        function=FunctionToolDefinition(
+            name=function_def.name,
+            description=function_def.description,
+            parameters=convert_gemini_function_param_to_mlflow_function_param(
+                function_def.parameters
+            ),
+        ),
+    )
+
+
+def convert_gemini_func_call_to_mlflow_tool_call(
+    func_call: "genai.protos.FunctionCall",
+) -> ToolCall:
+    """
+    Convert Gemini function call into MLflow's standard format (OpenAI compatible).
+    Ref: https://ai.google.dev/gemini-api/docs/function-calling
+
+    Args:
+        func_call: A genai.protos.FunctionCall object representing a single func call.
+
+    Returns:
+        ToolCall: MLflow's standard tool call object.
+    """
+    # original args object is not json serializable
+    args = func_call.args or {}
+
+    return ToolCall(
+        # Gemini does not have func call id
+        id=func_call.name,
+        type="function",
+        function=Function(name=func_call.name, arguments=json.dumps(dict(args))),
+    )
+
+
+def parse_gemini_content_to_mlflow_chat_messages(
+    content: "genai.content_types.ContentsType",
+) -> list[ChatMessage]:
+    """
+    Convert a gemini content to chat messages.
+
+    Args:
+        content: A genai.content_types.ContentsType object representing the model content.
+
+    Returns:
+        list[ChatMessage]: A list of MLflow's standard chat messages.
+    """
+    if isinstance(content, str):
+        # Assume str content is used only for user input
+        return [
+            ChatMessage(
+                role="user",
+                content=content,
+            )
+        ]
+    elif isinstance(content, list):
+        # either list of user inputs or multi-turn conversation
+        if not content:
+            return []
+        # when chat history is passed, parse content recursively
+        if hasattr(content[0], "parts"):
+            return sum(
+                [
+                    parse_gemini_content_to_mlflow_chat_messages(content_block)
+                    for content_block in content
+                ],
+                [],
+            )
+
+        # when multiple contents are passed by user
+        content_parts = []
+        for content_block in content:
```

Comment:
```suggestion
        for content_part in content:
```

nit: I'd use `content_part` here to distinguish this and the `content_block` used in L139-140. This one is a single part while the above is one level higher (a container of multiple parts).